### DISCUSSION TITLE
Should I use middleware to validate the response

### DISCUSSION THREAD

**Discussion by thedangler** [saloonphp/saloon#407] [Q&A]
Hello,
My responses from an endpoint are only valid if certain key values are set in the json response.
Using middleware, I'm thinking I can look for those key values and if they are not found some how set the response to 404? 
That way I do not need to look for the correct key values every time I want to use that end point. I've found myself repeating some of this logic.

The middleware docs mention not to alter the response. So how do you recommend I do this?

**Comment by Sammyjo20** [comment]
Hey @thedangler thanks for your patience! 

You can do it with middleware, but I would be inclined to extend Saloon's exception/failed request detection. On your connector (for global) or request you can extend a `hasRequestFailed` method and tell Saloon that it has failed without having to modify the status. If you are using the `AlwaysThrowOnError` trait too it would throw an exception too! 

See: https://docs.saloon.dev/the-basics/handling-failures#customising-when-saloon-thinks-a-request-has-failed

```php
public function hasRequestFailed(Response $response): ?bool
{
    if ($response->serverError() || $response->clientError()) {
         return true;
    }

    return str_contains($response->body(), 'Server Error');
}
```
